File: 185_KthLargestStream.cpp
/*
Problem: K-th Largest Element in Stream
Description: Design class to find kth largest element in stream.

Approach: Use min heap of size k. Top element is kth largest.
*/

#include <bits/stdc++.h>
using namespace std;

class KthLargest {
private:
    priority_queue<int, vector<int>, greater<int>> minHeap;
    int k;
    
public:
    KthLargest(int k, vector<int>& nums) : k(k) {
        for (int num : nums) {
            add(num);
        }
    }
    
    int add(int val) {
        if (minHeap.size() < k) {
            minHeap.push(val);
        } else if (val > minHeap.top()) {
            minHeap.pop();
            minHeap.push(val);
        }
        
        return minHeap.top();
    }
};

int main() {
    vector<int> nums = {4, 5, 8, 2};
    KthLargest kl(3, nums);
    
    cout << kl.add(3) << endl;   // Output: 4
    cout << kl.add(5) << endl;   // Output: 5
    cout << kl.add(10) << endl;  // Output: 5
    cout << kl.add(9) << endl;   // Output: 8
    cout << kl.add(4) << endl;   // Output: 8
    
    return 0;
}

/*
Time Complexity: add O(log k), constructor O(n log k)
Space Complexity: O(k)
*/

File: 185_KthLargestStream.py
"""
Problem: K-th Largest Element in Stream
Description: Design class to find kth largest element in stream.

Approach: Use min heap of size k. Top element is kth largest.
"""

from typing import List
import heapq

class KthLargest:
    def __init__(self, k: int, nums: List[int]):
        self.k = k
        self.minHeap = []
        for num in nums:
            self.add(num)
    
    def add(self, val: int) -> int:
        if len(self.minHeap) < self.k:
            heapq.heappush(self.minHeap, val)
        elif val > self.minHeap[0]:
            heapq.heappushpop(self.minHeap, val)
        
        return self.minHeap[0]

def main():
    nums = [4, 5, 8, 2]
    kl = KthLargest(3, nums)
    
    print(kl.add(3))   # Output: 4
    print(kl.add(5))   # Output: 5
    print(kl.add(10))  # Output: 5
    print(kl.add(9))   # Output: 8
    print(kl.add(4))   # Output: 8

if __name__ == "__main__":
    main()

"""
Time Complexity: add O(log k), constructor O(n log k)
Space Complexity: O(k)
"""
